Samples/WinMLSamplesGallery/TransformAsync/TransformAsync.cpp

```diff
@@ -0,0 +1,934 @@
+
+
+    // Schedule the inference task with the MF async work queue. Will call TransformAsync::Invoke when run.
+    RETURN_IF_FAILED(MFPutWorkItemEx(MFASYNC_CALLBACK_QUEUE_MULTITHREADED, result.get()));
+    return S_OK;
+}
+
+// SubmitEval returns an HRESULT to use to signal the async result's status
+HRESULT TransformAsync::SubmitEval(IMFSample* pInput)
+{
+    com_ptr<IMFSample> outputSample;
+    com_ptr<IMFSample> inputSample;
+    IDirect3DSurface src, dest;
+    inputSample.copy_from(pInput); 
+
+    // Select the next available model
+    DWORD dwCurrentSample = InterlockedIncrement(&m_sampleCounter);
+    m_modelIndex = (++m_modelIndex) % m_numThreads;
+    auto model = m_models[m_modelIndex].get();
+
+    //inputSample attributes to copy over to outputSample
+    LONGLONG duration = 0;
+    LONGLONG time = 0;
+    UINT64 markerID = 0;
+
+    TRACE((L"\n[Sample: %d | model: %d | ", currentSample, m_modelIndex));
+    TRACE((L" | SE Thread %d | ", std::hash<std::thread::id>()(std::this_thread::get_id())));
+
+     // Ensure we still have a valid d3d device
+    RETURN_IF_FAILED(CheckDX11Device());
+    // Ensure the allocator is set up 
+    RETURN_IF_FAILED(SetupAlloc());
+
+    // CheckDX11Device & SetupAlloc ensure we can now allocate a D3D-backed output sample. 
+    HRESULT hr = m_outputSampleAllocator->AllocateSample(outputSample.put());
+    if (FAILED(hr))
+    {
+        TRACE((L"Can't allocate any more samples, DROP"));
+    }
+
+    // We should have a sample now, whether or not we have a dx device
+    RETURN_IF_NULL_ALLOC(outputSample);
+
+    // Explicitly copy out pInput attributes, since copying IMFSamples is shallow. 
+    RETURN_IF_FAILED(inputSample->GetSampleDuration(&duration));
+    RETURN_IF_FAILED(inputSample->GetSampleTime(&time));
+    inputSample->GetUINT64(TransformAsync_MFSampleExtension_Marker, &markerID);
+
+    // Extract an IDirect3DSurface from the input and output samples to use for inference, 
+    src = SampleToD3Dsurface(inputSample.get());
+    dest = SampleToD3Dsurface(outputSample.get());
+
+    // Make sure this model isn't already running, if so drop the frame and move on. 
+    if(model->m_syncStarted == FALSE)
+    {
+        auto now = std::chrono::high_resolution_clock::now();
+        // Run model inference 
+        model->Run(src, dest); 
+        auto timePassed = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - now);
+        TRACE((L"Eval: %d", timePassed.count()));
+        src.Close();
+        dest.Close();
+        // Perform housekeeping out the output sample
+        RETURN_IF_FAILED(FinishEval(inputSample, outputSample, duration, time, markerID)); 
+    }
+    else {
+        TRACE((L"Model %d already running, DROP", m_modelIndex));
+    }
+
+    return S_OK;
+}
+
+// Clean up output sample and schedule MFHaveOutput task
+HRESULT TransformAsync::FinishEval(com_ptr<IMFSample> pInputSample, com_ptr<IMFSample> pOutputSample,
```

Comment:
>com_ptr<IMFSample> pOutputSample [](http://example.com/codeflow?start=68&length=32)

Unless a function needs the actual `ComPtr`/`com_ptr`, it's usually better to pass the interface rather than a specific wrapper of that interface. "be conservative in what you send, be liberal in what you accept"
[https://en.wikipedia.org/wiki/Robustness_principle](https://en.wikipedia.org/wiki/Robustness_principle)

Also, perfwise, it's just increments and decrements the atomic ref-count unnecessarily (perf death by a thousand cuts).

`IMFSample* pOutputSample`